### PR TITLE
rabbitmq_user module can change password without 'force: yes'

### LINKED ISSUE #3237: rabbitmq_user module doesn't change password without force
the rabbitmq_user account won't change the password of an existing user unless the force paramter is used. For example, this doesn't do what you would expect:

```
rabbitmq_user: user=guest password=newpasswordhere
```

On the other hand, this does the right thing:

```
rabbitmq_user: user=guest password=newpasswordhere force=yes
```

Ideally, we would fix rabbitmq_user, but at a minimum the docs should be updated to reflect this.

### PR DESCRIPTION
**Pull request by ar7z1**
##### SUMMARY

In this PR I fixed #3237.
Since RabbitMQ 3.6.5 ([release notes](https://rabbitmq.docs.pivotal.io/36/relnotes/release-notes-rabbitmq3.6.5.html)) `rabbitmqctl` contains `authenticate_user` option:
>rabbitmqctl authenticate_user is a new command that can be used to test user authentication.

I used this option to check user password.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

messaging/rabbitmq_user

##### ANSIBLE VERSION

```
ansible 2.4.0 (3237-change-rabbitmq-user-password-without-force 335fa4965c) last updated 2017/08/04 15:09:26 (GMT +300)
  config file = /home/art/.ansible.cfg
  configured module search path = [u'/home/art/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/art/projects/github/ansible/lib/ansible
  executable location = /home/art/projects/github/ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/messaging/rabbitmq_user.py [hunk 1]
```diff
@@ -229,6 +229,14 @@ def has_tags_modifications(self):
     def has_permissions_modifications(self):
         return sorted(self._permissions) != sorted(self.permissions)
 
+    def has_password_modification(self):
+        _, rc = self._exec(['authenticate_user', self.username, self.password], run_in_check_mode=True, check_rc=False)
+        # 'EX_NOUSER' (The user specified does not exist) from https://github.com/rabbitmq/rabbitmq-server/blob/master/include/rabbit_cli.hrl#L70
+        return rc == 67
```
Comment (ar7z1) [inline lib/ansible/modules/messaging/rabbitmq_user.py]: `rabbitmqctl authenticate_user` returns `67` ([EX_NOUSER](https://github.com/rabbitmq/rabbitmq-server/blob/master/include/rabbit_cli.hrl#L70)) when can't authenticate.
Old `rabbitmqctl` (RabbitMQ < 3.6.5) will return `64` ([EX_USAGE](https://github.com/rabbitmq/rabbitmq-server/blob/master/include/rabbit_cli.hrl#L68) exit code) for `authenticate_user` option.

File: lib/ansible/modules/messaging/rabbitmq_user.py [hunk 2]
```diff
@@ -31,8 +31,8 @@
   password:
     description:
       - Password of user to add.
-      - To change the password of an existing user, you must also specify
-        C(force=yes).
+      - To change the password of an existing user for RabbitMQ < 3.6.5,
+        you must also specify C(force=yes).
```
Comment (resmo) [inline lib/ansible/modules/messaging/rabbitmq_user.py]: Could you rephrase this a bit to show users this feature is available since 2.4.

Comment (ar7z1) [reply lib/ansible/modules/messaging/rabbitmq_user.py]: @resmo Yep. :-)
What do you think, if we write this:
> - To change the password of an existing user, when you use 
      ansible < 2.4 or RabbitMQ < 3.6.5, you must also specify C(force=yes).

Comment (resmo) [reply lib/ansible/modules/messaging/rabbitmq_user.py]: My suggestion would be:
~~~
- To change the password of an existing user, you must also specify C(force=yes).
- Since 2.4, C(force=yes) is optional with RabbitMQ 3.6.5 and later.
~~~

**Comment by ar7z1** [comment]
@resmo [Changed description](https://github.com/ansible/ansible/pull/27757/commits/c0e97add2f8f14d333cd33f841d21a6ab17e08cd).

File: lib/ansible/modules/messaging/rabbitmq_user.py (hunk 2 above)
Comment (ar7z1) [reply lib/ansible/modules/messaging/rabbitmq_user.py]: @resmo Ok, [changed](https://github.com/ansible/ansible/pull/27757/commits/c0e97add2f8f14d333cd33f841d21a6ab17e08cd) description.

**Comment by ar7z1** [comment]
@resmo Do I have to do something else in this PR?

**Comment by ar7z1** [comment]
@resmo I rebased on the latest devel and resolved merge conflicts.

**Comment by ar7z1** [comment]
@resmo Can you review the changes? ;-)

**Comment by resmo** [comment]
code LGTM

shipit

**Comment by resmo** [comment]
bot_status

**Comment by ar7z1** [comment]
@chrishoffman @manuel-sousa @romanek-adam Can you review this PR? ;-)

File: lib/ansible/modules/messaging/rabbitmq_user.py [hunk 3]
```diff
@@ -229,6 +229,14 @@ def has_tags_modifications(self):
     def has_permissions_modifications(self):
         return sorted(self._permissions) != sorted(self.permissions)
 
+    def has_password_modification(self):
```
Comment (resmo) [inline lib/ansible/modules/messaging/rabbitmq_user.py:232]: I reviewed the change again and I think a check for a none password is missing, because the password field is optional we should only validate the password if not None. 

~~~python
if self.password is None:
  return False
~~~

**Comment by resmo** [comment]
needs_revision

File: lib/ansible/modules/messaging/rabbitmq_user.py (hunk 3 above)
Comment (ar7z1) [reply lib/ansible/modules/messaging/rabbitmq_user.py:232]: @resmo Great catch! I'll change the PR. ;-)

Comment (ar7z1) [reply lib/ansible/modules/messaging/rabbitmq_user.py:232]: @resmo Fixed in 9831ab7.

**Comment by ar7z1** [comment]
ready_for_review

**Comment by resmo** [comment]
The password is optional (None) which as a user, I would expect the password just be left "as is". 

Clearing a password, I would implement like:

~~~yaml
- rabbitmq_user:
    user: joe
    password: ""
    state: present
~~~

**Comment by ar7z1** [comment]
@resmo Removed redundant `clear_password` in 4cdda0d.

**Comment by ar7z1** [comment]
ready_for_review

**Comment by krzysztof-magosa** [comment]
Hi,
Any update here?
Please let me know if you are going to resolve conflicts, otherwise I will resolve them and open new PR.
Kind Regards.

**Comment by ar7z1** [comment]
@krzysztof-magosa, I think, almost the same changes have already merged in #36280.
And we just need to close this PR.

**Comment by krzysztof-magosa** [comment]
Code from #36280 does not work.
It fails in line 191 because self._exec raises exception by default when rc != 0.

**Comment by ar7z1** [comment]
Hmm, ok, than I can fix it. Thank you!
